core/common/src/main/java/alluxio/network/connection/ThriftClientPool.java

```diff
@@ -0,0 +1,216 @@
+import alluxio.security.authentication.TransportProvider;
+import alluxio.thrift.AlluxioService;
+
+import org.apache.thrift.TException;
+import org.apache.thrift.protocol.TBinaryProtocol;
+import org.apache.thrift.protocol.TMultiplexedProtocol;
+import org.apache.thrift.protocol.TProtocol;
+import org.apache.thrift.transport.TTransport;
+import org.apache.thrift.transport.TTransportException;
+
+import java.io.IOException;
+import java.net.InetSocketAddress;
+import java.util.regex.Pattern;
+
+import javax.annotation.concurrent.ThreadSafe;
+
+/**
+ *  A pool to manage Alluxio thrift clients.
+ *
+ * @param <T> the Alluxio thrift service type
+ */
+@ThreadSafe
+public abstract class ThriftClientPool<T extends AlluxioService.Client>
+    extends DynamicResourcePool<T> {
+  private final TransportProvider mTransportProvider;
+  private final String mServiceName;
+  private final long mServiceVersion;
+  private final InetSocketAddress mAddress;
+  private final int mGcThresholdInSecs;
+
+  // Makes sure that the version is not checked for every new Client.
+  private volatile boolean mIsVersionChecked = false;
+
+  private static final int CONNECTION_OPEN_RETRY_BASE_SLEEP_MS = 50;
+  private static final int CONNECTION_OPEN_RETRY_MAX = 5;
+
+  /**
+   * The patterns of exception message when client and server transport frame sizes do not match
+   * or corrupted data (e.g. due to incorrect port configured).
+   */
+  private static final Pattern FRAME_SIZE_TOO_LARGE_EXCEPTION_PATTERN =
+      Pattern.compile("Frame size \\((\\d+)\\) larger than max length");
+  private static final Pattern FRAME_SIZE_NEGATIVE_EXCEPTION_PATTERN =
+      Pattern.compile("Read a negative frame size");
+
+  /**
+   * Creates a thrift client pool instance with a minimum capacity of 1.
+   *
+   * @param serviceName the service name (e.g. BlockWorkerClient)
+   * @param serviceVersion the service version
+   * @param address the server address
+   * @param maxCapacity the maximum capacity of the pool
+   * @param gcThresholdInSecs when a channel is older than this threshold and the pool's capacity
+   *        is above the minimum capaicty (1), it is closed and removed from the pool.
+   */
+  public ThriftClientPool(String serviceName, long serviceVersion, InetSocketAddress address,
+      int maxCapacity, int gcThresholdInSecs) {
+    super(Options.defaultOptions().setMaxCapacity(maxCapacity));
+    mTransportProvider = TransportProvider.Factory.create();
+    mServiceName = serviceName;
+    mServiceVersion = serviceVersion;
+    mAddress = address;
+    mGcThresholdInSecs = gcThresholdInSecs;
+  }
+
+  @Override
+  protected void closeResource(T client) {
+    // Note that the input and output protocol is the same in Alluxio.
+    TTransport transport = client.getOutputProtocol().getTransport();
+    if (transport.isOpen()) {
+      LOG.info("Closing thrift transport.");
+      transport.close();
+    } else {
+      LOG.info("Close a closed thrift transport.");
```

Comment:
I think it is ok because it should not be a frequent event to close a thrift transport with pooling logic. Let me know if you have a concern, i can change it.